### ISSUE TITLE
Segfault in is_typekey_ordered

### ISSUE THREAD

**Issue by simonster** [JuliaLang/julia#19504]
```
signal (11): Segmentation fault
while loading no file, in expression starting on line 0
is_typekey_ordered at /usr/local/julia-release-0.5/src/jltypes.c:1891
lookup_type at /usr/local/julia-release-0.5/src/jltypes.c:2020 [inlined]
inst_datatype at /usr/local/julia-release-0.5/src/jltypes.c:2184
jl_inst_concrete_tupletype_v at /usr/local/julia-release-0.5/src/jltypes.c:2378
arg_type_tuple at /usr/local/julia-release-0.5/src/gf.c:1128
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1925
#notify at ./<missing>:0
unknown function (ip: 0x7fbecc8cbe66)
jl_call_method_internal at /usr/local/julia-release-0.5/src/julia_internal.h:189 [inlined]
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1945
task_done_hook at ./task.jl:156
unknown function (ip: 0x7fbca9145982)
jl_call_method_internal at /usr/local/julia-release-0.5/src/julia_internal.h:189 [inlined]
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1945
jl_apply at /usr/local/julia-release-0.5/src/julia.h:1392 [inlined]
finish_task at /usr/local/julia-release-0.5/src/task.c:214 [inlined]
start_task at /usr/local/julia-release-0.5/src/task.c:261
unknown function (ip: 0xffffffffffffffff)
Allocations: 24810373 (Pool: 24808765; Big: 1608); GC: 46
fish: “julia” terminated by signal SIGSEGV (Address boundary error)
```

with

```
Julia Version 0.5.1-pre+31
Commit 6a1e339* (2016-11-17 17:50 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-3930K CPU @ 3.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, sandybridge)
```

I can try to reduce this if necessary, but it will be difficult, although it was consistently reproducible. I eventually determined that there was an undefined variable and the code should have thrown, but instead it segfaulted.

**Comment by yuyichao** [comment]
Code to reproduce?

**Comment by simonster** [comment]
There is code here: https://github.com/simonster/julia-issue-19504, but even modest tweaks appear to eliminate the bug. Running `fail.jl` reproduces the segfault for me on the system I on which I encountered this (with JLD 0.6.6 and JSON 0.8.0). I can't reproduce it on my OS X system.

**Comment by simonster** [comment]
This was probably fixed by the type system overhaul in 0.6.
